### PR TITLE
Check for the presence of the MySQL binaries before calling them

### LINKED ISSUE #388: Does not verify whether the MySQL client is installed
Bee has several commands that require having the MySQL client, which is not a system requirement for Backdrop. It can lead to frustrating errors, such as:

```
/var/www/bkdrp_proglo $ bee sqlc
sh: 1: mysql: not found
✓ gwolf@backdrop『33』/var/www/bkdrp_proglo $ bee db-drop
Are you sure you want to drop the contents of bkdrp_proglo ? (y/N): y
sh: 1: mysql: not found
sh: 1: mysql: not found

 [✔] The 'bkdrp_proglo' database was successfully dropped. 
 [✔] The 'bkdrp_proglo' database was successfully created. 
```

I will be submitting  a pull request proposing a patch that fixes this by checking before exec() / proc_open().

### PR DESCRIPTION
**Pull request by gwolf**
Fixes #388 
Patch to verify for the presence of `mysql` and `mysqldump` (and `gunzip`, as it was in the same command construction) before actually calling it.

Thanks!

### CODE REVIEW

File: commands/db.bee.inc [hunk 1]
```diff
@@ -348,3 +378,17 @@ function dbq_bee_callback($arguments, $options) {
     bee_message($e->getMessage(), 'error');
   }
 }
+
+/**
+ * Checks whether a command exists in the PATH, and if so, returns the full
+ * path. Returns null if no such command is available.
+ */
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Please follow the guidance for https://docs.backdropcms.org/doc-standards#functions
adding `@param` and `@return`

File: commands/db.bee.inc [hunk 2]
```diff
@@ -348,3 +378,17 @@ function dbq_bee_callback($arguments, $options) {
...
+function executable_for($cmd) {
+  foreach (explode(":", getenv('PATH')) as $dir) {
+    $candidate = "$dir/$cmd";
+    if (is_executable($candidate)) {
+      return $candidate;
+    }
+  }
+  return null;
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Please note
https://docs.backdropcms.org/php-standards#naming-conventions

> Constants should always be all-uppercase, with underscores to separate words. This includes pre-defined PHP constants like TRUE, FALSE, and NULL

```suggestion
  return NULL;
```

However, we don't normally use `NULL` as a return value. `FALSE` is simpler as the check for it in this case would be something like:
```php
if (!$cmd_bin) {
```

```suggestion
  return FALSE;
```

File: commands/db.bee.inc [hunk 3]
```diff
@@ -119,8 +119,14 @@ function db_export_bee_callback($arguments, $options) {
     $extra = '--no-tablespaces ';
   }
 
+  $export_command = executable_for('mysqldump');
+  if (is_null($export_command)) {
+    bee_message("The MySQL export command 'mysqldump' cannot be found in " .
+                "this system. Please install it and try again.", 'error');
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Within Backdrop, there is a function called `t()` that encloses translatable strings.  In bee, because we can't always rely on Backdrop being bootstrapped, we have our own function called `bt()` for translatable strings. Translatable strings are an exception (there are a couple of others) to the 80 character rule and the translatable string should be in a single string.

```suggestion
    bee_message(bt("The MySQL export command 'mysqldump' cannot be found in this system. Please install it and try again."), 'error');
```

File: commands/db.bee.inc [hunk 4]
```diff
@@ -119,8 +119,14 @@ function db_export_bee_callback($arguments, $options) {
...
+    return null;
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
    return FALSE;
```

File: commands/db.bee.inc [hunk 5]
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
   }
 
   // Import the database.
+  $mysql_bin = executable_for('mysql');
+  if (is_null($mysql_bin)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Above message about `bt()` applies

File: commands/db.bee.inc [hunk 6]
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
...
+    return null;
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
    return FALSE;
```

File: commands/db.bee.inc [hunk 7]
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
...
+  }
   $import_command = '';
   if ($gzip) {
-    $import_command .= "gunzip -c $filename | ";
+    $gunzip_bin = executable_for('gunzip');
+    if (is_null($gunzip_bin)) {
+      bee_message("The gzip decompressor command 'gunzip' cannot be found " .
+                  "in this system. Please install it and try again.", 'error');
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Above message about `bt()` applies

File: commands/db.bee.inc [hunk 8]
```diff
@@ -172,11 +178,23 @@ function db_import_bee_callback($arguments, $options) {
...
+      return null;
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
      return FALSE;
```

File: commands/db.bee.inc [hunk 9]
```diff
@@ -218,7 +236,13 @@ function db_drop_bee_callback($arguments, $options) {
   $connection_file = $connection_details['filename'];
 
   // Drop the existing backdrop database as configured.
-  $command = 'mysql ' . $connection_string . ' -e "drop database ' . $db_database . '";';
+  $mysql_command = executable_for('mysql');
+  if (is_null($mysql_command)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Above message about `bt()` applies

File: commands/db.bee.inc [hunk 10]
```diff
@@ -218,7 +236,13 @@ function db_drop_bee_callback($arguments, $options) {
...
+    return null;
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
    return FALSE;
```

File: commands/db.bee.inc [hunk 11]
```diff
@@ -265,7 +289,13 @@ function sql_bee_callback($arguments, $options) {
   $connection_file = $connection_details['filename'];
 
   // Open SQL command-line.
-  $command = 'mysql ' . $connection_string;
+  $mysql_command = executable_for('mysql');
+  if (is_null($mysql_command)) {
+    bee_message("The MySQL client command 'mysql' cannot be found in this " .
+                "system. Please install it and try again.", 'error');
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: Above message about `bt()` applies

File: commands/db.bee.inc [hunk 12]
```diff
@@ -265,7 +289,13 @@ function sql_bee_callback($arguments, $options) {
...
+    return null;
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
    return FALSE;
```

**Review by yorkshire-pudding — CHANGES_REQUESTED** [review]
Hi @gwolf 

I haven't yet tested this (I don't have any environments without these on to test).  I will run the automated tests again once the changes are made, but might also see if anyone else is able to test the negative conditions.  Most changes are about using `FALSE` instead of `null`, and making the `bee_message` strings translatable. Also a function docblock required.

Thank you

**Comment by gwolf** [comment]
Thanks!
I think I have addressed the needed changes for the coding standards. Sorry, I'm not used to PHP development, and of course, am completely new to this project.

**Comment by gwolf** [comment]
Ugh, the codesniffer still does not like my changes. And the details it gives are not really useful (just says "Error:" 😕)

**Comment by gwolf** [comment]
All tests passed! \o/

File: commands/db.bee.inc [hunk 13]
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
     bee_message($e->getMessage(), 'error');
   }
 }
+
+/**
+ * Checks whether a command exists in the PATH, and if so, returns the full
+ * path. Returns FALSE if no such command is available.
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: In line with my suggestion further down about moving this to `includes/filesystem.inc`, it needs a wider explanation of purpose to help other contributors or people writing their own commands.

```suggestion
 * Get the full path for a command executable if it exists in PATH.
 *
 * If your function calls an executable on the system rather than using `bee`,
 * `backdrop` or `php` functions, you can use this function to get the full
 * path of the executable or return FALSE if the executable doesn't exist. The
 * result can then be checked in your calling function, and you can exit with a
 * user friendly error message if the executable doesn't exist in the system.
```

File: commands/db.bee.inc [hunk 14]
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
...
+ *
+ * @param string $cmd
+ *   Name of the command to be searched.
+ *
+ * @return string
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
 * @return string|FALSE
```

File: commands/db.bee.inc [hunk 15]
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
...
+ *   Full path to the first occurence of $cmd in the search path.
+ *   Returns FALSE if no matching command is found.
+ */
+function executable_for($cmd) {
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]: I've been giving the name of this function a bit of thought, and I've come to the conclusion that this function actually belongs in `includes/filesystem.inc` (`includes` files are all included in `bee.php`) rather than in the `db.bee.inc` file.  Although only db functions currently use this in the core bee software, people writing their own commands or adding a bee command for their module might be grateful for it.  Please could you name the function `bee_get_executable_path` and move to  `includes/filesystem.inc`  - You'll need to update the calls to this function.

Also, can we have full names for variables please.

```suggestion
function bee_get_executable_path($command) {
```

File: commands/db.bee.inc [hunk 16]
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
...
+  foreach (explode(":", getenv('PATH')) as $dir) {
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
  foreach (explode(":", getenv('PATH')) as $directory) {
```

File: commands/db.bee.inc [hunk 17]
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
...
+    $candidate = "$dir/$cmd";
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
    $candidate = "$directory/$command";
```

File: commands/db.bee.inc [hunk 18]
```diff
@@ -348,3 +373,24 @@ function dbq_bee_callback($arguments, $options) {
     bee_message($e->getMessage(), 'error');
   }
 }
+
+/**
+ * Checks whether a command exists in the PATH, and if so, returns the full
+ * path. Returns FALSE if no such command is available.
+ *
+ * @param string $cmd
```
Comment (yorkshire-pudding) [inline commands/db.bee.inc]:
```suggestion
 * @param string $command
```

**Review by yorkshire-pudding — CHANGES_REQUESTED** [review]
Sorry @gwolf - a bit more to do.  After thinking about the name of the helper function, I've decided it would be better to go into `includes/filesystem.inc`.  There's also a few tweaks with the function doc bloc and variable names.

File: commands/db.bee.inc (hunk 13 above)
Comment (gwolf) [reply commands/db.bee.inc]: Completely makes sense. I have moved the function to the suggested file (and renamed to a more general, project-friendly way).

**Comment by yorkshire-pudding** [comment]
Thanks @gwolf - this is looking good. I tested the negative condition by temporarily changing the defined executable in the `db_export_bee_callback()` function so it couldn't be found.  The function works well.

Thank you for your patience in getting this over the line.
